controllers/etcd_controller.go

```diff
@@ -686,21 +703,109 @@ func (r *EtcdReconciler) reconcileEtcd(ctx context.Context, logger logr.Logger,
-		deployWaiter = gardenercomponent.OpWaiter(stsDeployer)
+		stsDeployer       = componentsts.New(r.Client, logger, statefulSetValues)
+		deployWaiter      = gardenercomponent.OpWaiter(stsDeployer)
+		peerUrlTLSEnabled *bool
 	)
 
 	if err := deployWaiter.Deploy(ctx); err != nil {
-		return nil, nil, err
+		return reconcileResult{}, err
+	}
+
+	// This needs to be done because component.Deploy cannot return a result today
+	if err == nil {
+		peerUrlTLSEnabled = pointer.BoolPtr(stsDeployer.IsPeerUrlTLSChangedToEnabled())
 	}
 
 	sts, err := stsDeployer.Get(ctx)
+	return reconcileResult{
+		peerUrlTLSEnabled: peerUrlTLSEnabled,
+		svcName:           &serviceValues.ClientServiceName,
+		sts:               sts,
+	}, err
+}
+
+// -------------------------- BEGIN: To be removed later --------------------------------------
+
+// capturePeerUrlTLSEnabledStatus captures the state of peer URL TLS enablement in etcd status.
+// It will fetch any existing ConfigMap and will derive the current state of PeerTLS enablement.
+// This method is a temporary arrangement made for https://github.com/gardener/etcd-druid/issues/416.
+// Once etcd status has captures the peer TLS information once then it will be safe to remove this method.
+func (r *EtcdReconciler) capturePeerUrlTLSEnabledStatus(ctx context.Context, etcd *druidv1alpha1.Etcd, configMapNamespacedName types.NamespacedName) error {
+	// Only if the etcd status does not contain PeerUrlTLSEnabled then look it up from the existing ConfigMap and initialize it.
+	// If it is already existing in the etcd status then at the end of the successful reconciliation run it will get updated to the new value
+	peerUrlTLSEnabledFromCm, err := r.getPeerUrlTLSEnabledStatusFromConfigMap(ctx, configMapNamespacedName)
+	if err != nil {
+		return err
+	}
+	// Only if the original etcd status did not have PeerUrlTLSEnabled set it should be set to the current status from ConfigMap.
+	// Any further updates will be done once the changes are applied to StatefulSet after etcd reconciliation is over.
+	if peerUrlTLSEnabledFromCm != nil && peerUrlTLSEnabledFromCm != etcd.Status.PeerUrlTLSEnabled {
+		return controllerutils.TryUpdateStatus(ctx, retry.DefaultBackoff, r.Client, etcd, func() error {
+			etcd.Status.PeerUrlTLSEnabled = peerUrlTLSEnabledFromCm
+			r.logger.Info("Patching etcd, setting PeerUrlTLSEnabled", "name", etcd.Name, "peerUrlTLSEnabledFromCm", peerUrlTLSEnabledFromCm)
+			return nil
+		})
+	}
+	return nil
+}
 
-	return &serviceValues.ClientServiceName, sts, err
+func (r *EtcdReconciler) getPeerUrlTLSEnabledStatusFromConfigMap(ctx context.Context, configMapNamespacedName types.NamespacedName) (*bool, error) {
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configMapNamespacedName.Name,
+			Namespace: configMapNamespacedName.Namespace,
+		},
+	}
+	err := r.Client.Get(ctx, configMapNamespacedName, cm)
+	switch {
+	case apierrors.IsNotFound(err):
+		return nil, nil
```

Comment:
It was an open point and my initial assumption was that if the config map is not there then i cannot assume any value for peerUrlTLS and therefore just return nil with the hope that it will get created. But I think this assumption is true in the gardener context and if we see usage of etcd-druid sans gardener then maybe it is better to just return the error and not ignore it.

Reply:
Got it, thank. Why don't we call `capturePeerUrlTLSEnabledStatus` later in the flow, i.e. right before the `StatefulSet` is deployed? We don't even need to consult the `ConfigMap`, but we can enhance the ConfigMap-Component as it already has the data in memory.

Reply:
The reason to always depend upon ConfigMap was that in the event of a crash, the in-memory state is not lost. Can you elaborate of what you have in mind?